# 🧪 Prueba Técnica NLP - Análisis Exploratorio y Preprocesamiento

👋 ¡Bienvenido(a) a esta prueba técnica!

En este ejercicio abordaremos el análisis de un conjunto de datos ampliamente utilizado en el ámbito del procesamiento de lenguaje natural: el CNN/DailyMail News Dataset. Este dataset contiene miles de artículos de noticias reales junto con sus respectivos resúmenes editoriales. La tarea está orientada a demostrar habilidades en análisis exploratorio de datos (EDA), preprocesamiento de texto y comprensión de datos no estructurados.

Esta prueba busca evaluar tanto la capacidad técnica como el pensamiento analítico aplicado al manejo de texto, una habilidad clave en proyectos de inteligencia artificial, sistemas de recomendación, motores de búsqueda, resúmenes automáticos y asistentes conversacionales.

#  🎯 Objetivo

El objetivo principal de esta prueba es realizar un análisis exploratorio y preprocesamiento textual del dataset CNN/DailyMail, con el fin de:

* Comprender la estructura, distribución y calidad de los textos disponibles.
* Aplicar técnicas de limpieza, normalización y análisis estadístico del lenguaje.

* Identificar patrones relevantes en los artículos y sus resúmenes, evaluando la viabilidad de aplicar modelos de NLP para tareas como resumen automático o clasificación temática.

Este ejercicio busca simular un escenario real en el que es necesario preparar datos textuales para su posterior uso en modelos de lenguaje o sistemas inteligentes.

In [19]:
# Install the latest versions of datasets and huggingface_hub
!pip install --upgrade datasets huggingface_hub fsspec

  Using cached fsspec-2025.3.2-py3-none-any.whl.metadata (11 kB)


In [33]:
from datasets import load_dataset

dataset = load_dataset("cnn_dailymail", "3.0.0")
df = dataset['train'].to_pandas()


In [21]:
#df.head(10)

# Solución

## Comprensión de los datos

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 287113 entries, 0 to 287112
Data columns (total 3 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   article     287113 non-null  object
 1   highlights  287113 non-null  object
 2   id          287113 non-null  object
dtypes: object(3)
memory usage: 6.6+ MB


In [23]:
df['id'].duplicated().any()

False

In [24]:
import pandas as pd

# Mostrar el contenido completo de cada columna
pd.set_option('display.max_colwidth', None)

#df.head(10)

In [25]:
# Longitud de cada texto en la columna "article"
longitudes = df["article"].str.len()

# Máx
longitud_maxima = longitudes.max()
# mín
longitud_minima = longitudes.min()

print(f"Longitud máxima: {longitud_maxima}")
print(f"Longitud mínima: {longitud_minima}")

Longitud máxima: 15925
Longitud mínima: 48


Conclusiones:
El DataFrame analizado contiene tres variables de tipo string, cada una con 287,113 filas. No se identificaron datos duplicados ni faltantes. La clave primaria no presenta anomalías y se encuentra en buen estado. No obstante, las otras dos variables sí presentan inconsistencias. La columna "highlights" contiene el carácter "\n" en todas las filas, lo cual no es adecuado y debería corregirse. Por otro lado, la columna de artículos incluye símbolos como "--", que también deberían ser eliminados para garantizar la calidad de los datos. Otro aspecto relevante es la presencia recurrente de la frase "E-mail to a friend" casi al final o al final de los artículos; esta frase, junto con el contenido que le sigue, no aporta valor y resulta redundante

 La extensión de los artículos varía considerablemente, con longitudes que oscilan entre 15,925 caracteres en el caso más extenso y 48 caracteres en el más corto. Esta amplia dispersión sugiere que la columna "highlights", que corresponde al resumen de los artículos, también presenta una variabilidad significativa en su longitud.

## Limpieza y Transformación

### Variable "highlights"

In [34]:
# Variable "highlights"
df['highlights'] = df['highlights'].str.replace('\n', ' ', regex=True) #Elimina "\n"
df['highlights'] = df['highlights'].str.replace(r'\s+\.', '.', regex=True) #Elimina espacios innecesarios
df['highlights'] = df['highlights'].str.replace('NEW:', '', regex=False) #Elimina "NEW:"

#df.head(10)


### Variable "article"

In [ ]:
df['article'] = df['article'].str.replace(r'E-mail to a friend.*', '', regex=True) #Elimina "E-mail to a friend*" y su continuación
df.head(15)

article  \
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  LONDON, England (Reuters) -- Harry Potter star Daniel Radcliffe gains access to a reported £20 million ($41.1 million) fortune as he turns 18 on Monday, but he insists the money won't cast a spell on him. Daniel Radcliffe as Harry Potter in "Harry Potter and the Order of the Phoenix" To the disappointment of gossip columnists around the world, the young actor says he has no plans to fritter his cash away on fast cars, drink and celebrity parties. "I don't plan to be one of those people who, as soon as they turn 18, suddenly buy themselves a massive sports car collection or something similar," he told an Australian interviewer earlier this month. "I don't think I'll 

In [38]:
df[['new', 'article2']] = df['article'].str.split(r'--', n=1, expand=True) #Separa la columna "article" por el delimitador "--"
df.head(10)

article  \
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 LONDON, England (Reuters) -- Harry Potter star Daniel Radcliffe gains access to a reported £20 million ($41.1 million) fortune as he turns 18 on Monday, but he insists the money won't cast a spell on him. Daniel Radcliffe as Harry Potter in "Harry Potter and the Order of the Phoenix" To the disappointment of gossip columnists around the world, the young actor says he has no plans to fritter his cash away on fast cars, drink and celebrity parties. "I don't plan to be one of those people who, as soon as they turn 18, suddenly buy themselves a massive sports car collection or something similar," he told an Australian interviewer earlier this month. "I don't think I'll b

In [39]:
df.drop(["article"], axis=1, inplace=True) #Elimina la variable "Article"
#df.head(10)

### Variable article2

In [40]:
df['article2'] = df['article2'].str.replace('--', '', regex=True) #Elimina "--" de la columna 'article2'
df.head(5)

highlights  \
0                                                                    Harry Potter star Daniel Radcliffe gets £20M fortune as he turns 18 Monday. Young actor says he has no plans to fritter his cash away. Radcliffe's earnings from first five Potter films have been held in trust fund.   
1  Mentally ill inmates in Miami are housed on the "forgotten floor" Judge Steven Leifman says most are there as a result of "avoidable felonies" While CNN tours facility, patient shouts: "I am the son of the president" Leifman says the system is unjust and he's fighting for change.   
2                                                                  "I thought I was going to die," driver says. Man says pickup truck was folded in half; he just has cut on face. Driver: "I probably had a 30-, 35-foot free fall" Minnesota bridge collapsed during rush hour Wednesday.   
3                                                                                                    Five small polyps found during procedure; "none worrisome," spokesman says. President reclaims powers transferred to vice president. Bush undergoes routine colonoscopy at Camp David.   
4                  NFL chief, Atlanta Falcons owner critical of Michael Vick's conduct. NFL suspends Falcons quarterback indefinitely without pay. Vick admits funding dogfighting operation but says he did not gamble. Vick due in federal court Monday; future in NFL remains uncertain.   

                                         id  \
0  42c027e4ff9730fbb3de84c1af0d2c506e41c3e4   
1  ee8871b15c50d0db17b0179a6d2beab35065f1e9   
2  06352019a19ae31e527f37f7571c6dd7f0c5da37   
3  24521a2abb2e1f5e34e6824e0f9e56904a2b0e88   
4  7fe70cc8b12fab2d0a258fababf7d9c6b5e1262a   

                                                                                                                                                                                                                                                                                                                                                                                     new  \
0                                                                                                                                                                                                                                                                                                                                                             LONDON, England (Reuters)    
1  Editor's note: In our Behind the Scenes series, CNN correspondents share their experiences in covering news and analyze the stories behind the events. Here, Soledad O'Brien takes users inside a jail where many of the inmates are mentally ill. An inmate housed on the "forgotten floor," where many mentally ill inmates are housed in Miami before trial. MIAMI, Florida (CNN)    
2                                                                                                                                                                                                                                                                                                                                                          MINNEAPOLIS, Minnesota (CNN)    
3                                                                                                                                                                                                                                                                                                                                                                      WASHINGTON (CNN)    
4                                                                                                                                                                                                                                                                                                                                                                              

### Variable "new"

In [ ]:
df.tail(2) #Muestra las dos ultimas filas


highlights  \
287111  The Black Country Ale Tairsters have been to 16,337 pubs, hotels and wine bars. They fill in an elaborate score card and leave a personalised beer mat in every bar. Women are allowed to the pub but not on boozy weekends away.   
287112             Facebook page supporting tradition gains one million 'likes' in a day. 'Don't let the Netherlands' most beautiful tradition disappear,' it says. UN has condemned the tradition claiming it reflects racial prejudice.   

                                              id  \
287111  32a1f9e5c37a938c0c0bca1a1559247b9c4334b2   
287112  8ec9ff4d633dd4cc26d53f503c33f7464b43c36e   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          

In [ ]:
df.loc[df['article2'].isna(), 'article2'] = df['new'] #Me rellena los nulos de "article2" con las filas de "new"
df.loc[df['article2'] == df['new'], 'new'] = None #Me elimina las filas de "new" que sean similares a "article2"
df.tail(2)

highlights  \
287111  The Black Country Ale Tairsters have been to 16,337 pubs, hotels and wine bars. They fill in an elaborate score card and leave a personalised beer mat in every bar. Women are allowed to the pub but not on boozy weekends away.   
287112             Facebook page supporting tradition gains one million 'likes' in a day. 'Don't let the Netherlands' most beautiful tradition disappear,' it says. UN has condemned the tradition claiming it reflects racial prejudice.   

                                              id   new  \
287111  32a1f9e5c37a938c0c0bca1a1559247b9c4334b2  None   
287112  8ec9ff4d633dd4cc26d53f503c33f7464b43c36e  None   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        

In [48]:
df.head(5)

highlights  \
0                                                                    Harry Potter star Daniel Radcliffe gets £20M fortune as he turns 18 Monday. Young actor says he has no plans to fritter his cash away. Radcliffe's earnings from first five Potter films have been held in trust fund.   
1  Mentally ill inmates in Miami are housed on the "forgotten floor" Judge Steven Leifman says most are there as a result of "avoidable felonies" While CNN tours facility, patient shouts: "I am the son of the president" Leifman says the system is unjust and he's fighting for change.   
2                                                                  "I thought I was going to die," driver says. Man says pickup truck was folded in half; he just has cut on face. Driver: "I probably had a 30-, 35-foot free fall" Minnesota bridge collapsed during rush hour Wednesday.   
3                                                                                                    Five small polyps found during procedure; "none worrisome," spokesman says. President reclaims powers transferred to vice president. Bush undergoes routine colonoscopy at Camp David.   
4                  NFL chief, Atlanta Falcons owner critical of Michael Vick's conduct. NFL suspends Falcons quarterback indefinitely without pay. Vick admits funding dogfighting operation but says he did not gamble. Vick due in federal court Monday; future in NFL remains uncertain.   

                                         id  \
0  42c027e4ff9730fbb3de84c1af0d2c506e41c3e4   
1  ee8871b15c50d0db17b0179a6d2beab35065f1e9   
2  06352019a19ae31e527f37f7571c6dd7f0c5da37   
3  24521a2abb2e1f5e34e6824e0f9e56904a2b0e88   
4  7fe70cc8b12fab2d0a258fababf7d9c6b5e1262a   

                                                                                                                                                                                                                                                                                                                                                                                     new  \
0                                                                                                                                                                                                                                                                                                                                                             LONDON, England (Reuters)    
1  Editor's note: In our Behind the Scenes series, CNN correspondents share their experiences in covering news and analyze the stories behind the events. Here, Soledad O'Brien takes users inside a jail where many of the inmates are mentally ill. An inmate housed on the "forgotten floor," where many mentally ill inmates are housed in Miami before trial. MIAMI, Florida (CNN)    
2                                                                                                                                                                                                                                                                                                                                                          MINNEAPOLIS, Minnesota (CNN)    
3                                                                                                                                                                                                                                                                                                                                                                      WASHINGTON (CNN)    
4                                                                                                                                                                                                                                                                                                                                                                              

In [52]:
df['journal'] = df['new'].str.extract(r'\((.*?)\)')
df.head(3)

highlights  \
0                                                                    Harry Potter star Daniel Radcliffe gets £20M fortune as he turns 18 Monday. Young actor says he has no plans to fritter his cash away. Radcliffe's earnings from first five Potter films have been held in trust fund.   
1  Mentally ill inmates in Miami are housed on the "forgotten floor" Judge Steven Leifman says most are there as a result of "avoidable felonies" While CNN tours facility, patient shouts: "I am the son of the president" Leifman says the system is unjust and he's fighting for change.   
2                                                                  "I thought I was going to die," driver says. Man says pickup truck was folded in half; he just has cut on face. Driver: "I probably had a 30-, 35-foot free fall" Minnesota bridge collapsed during rush hour Wednesday.   

                                         id  \
0  42c027e4ff9730fbb3de84c1af0d2c506e41c3e4   
1  ee8871b15c50d0db17b0179a6d2beab35065f1e9   
2  06352019a19ae31e527f37f7571c6dd7f0c5da37   

                                                                                                                                                                                                                                                                                                                                                                                     new  \
0                                                                                                                                                                                                                                                                                                                                                             LONDON, England (Reuters)    
1  Editor's note: In our Behind the Scenes series, CNN correspondents share their experiences in covering news and analyze the stories behind the events. Here, Soledad O'Brien takes users inside a jail where many of the inmates are mentally ill. An inmate housed on the "forgotten floor," where many mentally ill inmates are housed in Miami before trial. MIAMI, Florida (CNN)    
2                                                                                                                                                                                                                                                                                                                                                          MINNEAPOLIS, Minnesota (CNN)    

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      